### PR TITLE
[cmd/mdatagen] Enhance mdatagen to support string parsing in RecordXDataPoint functions

### PR DESCRIPTION
**Pull request by dehaansa**
**Description:** Add support to mdatagen that will parse a string to integer/float before creating the data point in the RecordXDataPoint function if the input_type is set in metadatayaml.

This was step 1/simple version of the linked issue, did not seek to also support interface{} parsing

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/7574

**Testing:** Function utilized in apache & mysql, passes tests.

### CODE REVIEW

File: cmd/mdatagen/metrics_v2.tmpl [hunk 1]
```diff
@@ -188,6 +195,15 @@ func (mb *MetricsBuilder) Emit(ro ...ResourceOption) pdata.Metrics {
...
 
+func logFailedParse(logger *zap.Logger, expectedType, metric, value string) {
+	logger.Info(
+		"failed to parse value",
+		zap.String("expectedType", expectedType),
+		zap.String("metric", metric),
+		zap.String("value", value),
+	)
+}
+
```
Comment (djaglowski) [inline cmd/mdatagen/metrics_v2.tmpl:206]: I don't think we want to both log and return the error.

**Comment by dmitryax** [comment]
I think adding too many functions that are not supposed to be used seems redundant and may confuse the metrics builder users. 

What if we add another field for metrics in metadata yaml like `input_type`?

For example:

```yaml
  process.memory.virtual_usage:
    enabled: true
    description: Virtual memory size.
    unit: By
    sum:
      value_type: int
      input_type: string
      aggregation: cumulative
      monotonic: false
```

If `input_type` is not provided, we render a regular function:

```golang
func (mb *MetricsBuilder) RecordProcessMemoryVirtualUsageDataPoint(ts pdata.Timestamp, val int64) {
```

If `input_type` set to `string`, we render:

```golang
func (mb *MetricsBuilder) RecordProcessMemoryVirtualUsageDataPoint(ts pdata.Timestamp, val string) error {
```

cc @djaglowski

**Comment by djaglowski** [comment]
@dmitryax I think it's a good idea. +1

**Comment by dehaansa** [comment]
@djaglowski @dmitryax I believe I've reimplemented based on your feedback, let me know how it looks!

File: cmd/mdatagen/metrics_v2.tmpl [hunk 2]
```diff
@@ -198,11 +202,39 @@ func (mb *MetricsBuilder) Emit(ro ...ResourceOption) pmetric.Metrics {
...
 func (mb *MetricsBuilder) Record{{ $name.Render }}DataPoint(ts pcommon.Timestamp
-	{{- if $metric.Data.HasMetricValueType }}, val {{ $metric.Data.MetricValueType.BasicType }}{{ end }}
-	{{- range $metric.Attributes -}} , {{ .RenderUnexported }}AttributeValue string{{ end }}) {
+	{{- if $metric.Data.HasMetricInputType }}, val {{ $metric.Data.MetricInputType.String }}
+	{{- else }}
+	{{- if $metric.Data.HasMetricValueType }}, val {{ $metric.Data.MetricValueType.BasicType }}{{- end }}
+	{{- end -}}
+	{{- range $metric.Attributes -}} , {{ .RenderUnexported }}AttributeValue string{{ end }})
+	{{- if $metric.Data.HasMetricInputType }} error{{ end }} {
+	{{- if $metric.Data.HasMetricInputType }}
```
Comment (djaglowski) [inline cmd/mdatagen/metrics_v2.tmpl:211]:
```suggestion
	{{- if $metric.Data.HasMetricInputType }} error{{ end }} {
	{{- if $metric.Data.HasMetricInputType }}
```
Should these be combined?

**Review by djaglowski — APPROVED** [review]
LGTM

File: cmd/mdatagen/metrics_v2.tmpl (hunk 2 above)
Comment (dehaansa) [reply cmd/mdatagen/metrics_v2.tmpl:211]: I couldn't think of a very clean way to do that with the single `{` at the end of the line there that's needed either way for the end of the function declaration

**Review by dmitryax — APPROVED** [review]

**Pull request merged**